### DISCUSSION TITLE
Stricter version of EmptyCatchBlock?

### DISCUSSION THREAD

**Discussion by maikelsteneker** [pmd/pmd#3095] [Ideas]
PMD includes the rule EmptyCatchBlock that flags empty catch blocks. The idea behind this is that exceptions shouldn't be ignored, but should be acted on or reported.

Now consider code such as the following:
```
try {
    calculateChecksum(filename);
} catch (Exception ex) {
    LOGGER.error("Error calculating checksum for file {}", filename);
}
```

In this case, the code can probably be improved by catching a more specific error than the generic `Exception`, but in addition to this, it would probably be useful to keep the message or stacktrace for logging purposes. I would formalize this to: **the exception variable should be referred to somewhere in the catch block**. This could be a separate rule, or it could be an addition to an existing PMD rule.

Any thoughts on this? Should I create an issue for this?

**Comment by oowekyala** [comment]
Isn't UnusedLocalVariable exactly that? It isn't specific to catch blocks, but, it doesn't need to be.

**Reply by maikelsteneker** [reply]
Thank you for your reply. You are indeed completely correct! But UnusedLocalVariable doesn't trigger for me on this code. I guess this is a false positive?

**Reply by oowekyala** [reply]
Hey, sorry for the late feedback. Actually UnusedLocalVariable only cares for local variable declarations (not other kinds of "locals", like method formals, resources, or exception parameters). In PMD 7 these will be probably be merged, see #2972

**Reply by maikelsteneker** [reply]
No problem, I'm happy to see a reply :) So if I can summarize: currently, PMD includes a number of rules ([UnusedLocalVariable](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedlocalvariable), [UnusedPrivateField](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedprivatefield), [UnusedFormalParameter](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedformalparameter), and [UnusedPrivateMethod](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod)) which all detect unused "thing" in one context or another. Currently, there is no rule for exception parameters as well as some other elements. For PMD 7, these rules will be merged, presumably with other declarations as well. I'll keep following #2972 for updates on this!

**Reply by oowekyala** [reply]
Exactly that :+1:
